go/host/enclave/guardian.go

```diff
@@ -515,45 +517,81 @@ func (g *Guardian) periodicBatchProduction() {
-	if interval == 0 {
-		interval = 3 * time.Second
+	// rollup at least at every g.rollupInterval
+	rollupInterval := g.rollupInterval
+	if rollupInterval == 0 {
+		g.logger.Crit("invalid rollup interval 0")
 	}
-	rollupTicker := time.NewTicker(interval)
-	// attempt to produce rollup every time the timer ticks until we are stopped/interrupted
+	rollupIntervalBasedTicker := time.NewTicker(rollupInterval)
+
+	// every 5sec check if it should create a new rollup
+	rollupSizeCheckTicker := time.NewTicker(5 * time.Second)
+
 	for {
-		if g.hostInterrupter.IsStopping() {
-			rollupTicker.Stop()
-			return // stop periodic rollup production
-		}
 		select {
-		case <-rollupTicker.C:
+		case <-rollupIntervalBasedTicker.C:
 			if !g.state.IsUpToDate() {
 				// if we're behind the L1, we don't want to produce rollups
 				g.logger.Debug("skipping rollup production because L1 is not up to date", "state", g.state)
 				continue
 			}
-			lastBatchNo, err := g.sl.L1Publisher().FetchLatestSeqNo()
+
+			fromBatch, err := g.getLastestBatchNo()
 			if err != nil {
 				g.logger.Warn("encountered error while trying to retrieve latest sequence number", log.ErrKey, err)
 				continue
 			}
-			fromBatch := lastBatchNo.Uint64()
-			if lastBatchNo.Uint64() > common.L2GenesisSeqNo {
-				fromBatch++
-			}
+
 			producedRollup, err := g.enclaveClient.CreateRollup(fromBatch)
 			if err != nil {
 				g.logger.Error("unable to produce rollup", log.ErrKey, err)
 			} else {
+				fmt.Println(time.Now().String(), " - PRODUCING rollup from config'd rollupInterval - lastBatchNo: ", fromBatch)
 				g.sl.L1Publisher().PublishRollup(producedRollup)
+				fmt.Println(time.Now().String(), " - FINISHED PRODUCED rollup from config'd rollupInterval ")
+			}
+
+		case <-rollupSizeCheckTicker.C:
+			if !g.state.IsUpToDate() {
+				// if we're behind the L1, we don't want to produce rollups
+				g.logger.Debug("skipping rollup production because L1 is not up to date", "state", g.state)
+				continue
+			}
+
+			fromBatch, err := g.getLastestBatchNo()
+			if err != nil {
+				g.logger.Warn("encountered error while trying to retrieve latest sequence number", log.ErrKey, err)
+				continue
 			}
+
+			fmt.Println(time.Now().String(), " - 5sec rollup creation CHECK - lastBatchNo: ", fromBatch)
+			availBatchesSumSize, err := g.getBatchesAfterSize(fromBatch)
+			if err != nil {
+				g.logger.Error("unable to GetBatchesAfterSize rollup", log.ErrKey, err)
+				continue
+			}
+
+			fmt.Println(time.Now().String(), " - Sum of bytes of Batches not yet rolled up: - ", availBatchesSumSize, " bytes")
+			if availBatchesSumSize >= g.maxRollupSize {
+				producedRollup, err := g.enclaveClient.CreateRollup(fromBatch)
+				if err != nil {
+					g.logger.Error("unable to produce rollup", log.ErrKey, err)
+				} else {
+					fmt.Println(time.Now().String(), " - PRODUCING rollup from 5sec check ")
+					g.sl.L1Publisher().PublishRollup(producedRollup)
+				}
+				rollupSizeCheckTicker.Reset(5 * time.Second)
```

Comment:
these should be configs

Reply:
Yeah, the L1 blocktime should be available in the host now, could tie it to that.

I'd say if you end up doing the change  I suggested above of using a single timer then you could replace the reset with a 1 block *minimum* time since last rollup. Like pseudocode:

```
if time.Since(lastSuccessfulRollup) < l1BlockTime {
   continue // only just produced rollup, allow enclave time to consume
}
if time.Since(lastSuccessfulRollup) > rollupInterval || sizeTooBig() {
   doTheRollup()
}
```